### PR TITLE
Fire alarms now monitor atmospheric conditions, updating every 4 seconds.

### PR DESCRIPTION
**Pull request by F-OS**
[FEATURE] [BALANCE] [WIP]
## About the PR
This PR causes fire alarms to trip:
1. if pressure on the tile where the fire alarm is installed drops below 50 PSI
2. if the number of moles of plasma falls above 5 mol on the tile.
3. if water gets above levels where it would drown people.

## Why's this needed? 
This might be helpful in preventing both new players walking into dangerous situations and super-large breaches, and also make zero-effort bombings more difficult. In particular, this would make it so if you want to bomb, for instance, cog 1's halls, you would need to disable fire alarms first.

## Changelog
```
(u)Fosstar:
(*) Fire alarms now watch for flooding, plasma, and low pressure, along with fire.
```

### CODE REVIEW

**Comment by F-OS** [comment]
note that this is currently very fucked but I hope the code I have so far shows my general idea

**Comment by F-OS** [comment]
alright I'm sorry about all the commits my git tree is fucked 6 ways to sunday god fucking damnit aaa

**Comment by Kyle2143** [comment]
> if pressure on the tile where the fire alarm is installed drops below 50 PSI

Fire alarms used to do this on oshan to prevent water breaches. It was awful. It really did more to kill players than prevent them from dying.

**Comment by MarkNais** [comment]
We spoke in Discord.
I agree with Kyle: No to flood-detection.
The alarm unit could use a 3 to 5 second cooldown after a manual dis-arm before it "wakes up" again. Would allow for trapped crew to escape.

File: code/mob/living/silicon/ai.dm [hunk 1]
```diff
@@ -684,7 +685,13 @@ var/list/ai_emotions = list("Happy" = "ai_happy",\
...
 	if (isdead(src))
 		return 1
-	var/list/L = src.alarms[class]
+	var/indexclass = class
+	for(var/loopclass in list("Fire", "Low Pressure", "Flammable Atmosphere", "Flood", "Manual Trip"))
+		if(class == loopclass)
+			indexclass = "Fire"
+			break
+
+	var/list/L = src.alarms[indexclass]
```
Comment (MarkNais) [inline code/mob/living/silicon/ai.dm:694]: This block of code isn't doing anything?

File: code/mob/living/silicon/robot.dm [hunk 2]
```diff
@@ -965,7 +965,13 @@
...
 		if (isdead(src))
 			return 1
-		var/list/L = src.alarms[class]
+		var/indexclass = class
+		for(var/loopclass in list("Fire", "Low Pressure", "Flammable Atmosphere", "Flood", "Manual Trip"))
+			if(class == loopclass)
+				indexclass = "Fire"
+				break
+
+		var/list/L = src.alarms[indexclass]
```
Comment (MarkNais) [inline code/mob/living/silicon/robot.dm:974]: This block of code also isn't doing anything?

File: code/obj/machinery/computer/atmosphere.dm [hunk 3]
```diff
@@ -112,7 +112,12 @@ Atmos alert computer
 /obj/machinery/computer/atmosphere/alerts/proc/triggerAlarm(var/class, area/A, var/O, var/alarmsource)
 	if(status & (BROKEN|NOPOWER))
 		return
-	var/list/L = src.alarms[class]
+
+	var/indexclass = class
+	if(class in list("Fire", "Low Pressure", "Flammable Atmosphere", "Flood", "Manual Trip"))
+		indexclass = "Fire"
+	var/list/L = src.alarms[indexclass]
+
```
Comment (MarkNais) [inline code/obj/machinery/computer/atmosphere.dm:120]: Always make the alarm type Fire?

File: code/obj/machinery/firealarm.dm [hunk 4]
```diff
@@ -33,7 +34,7 @@
 
 	if(!net_id)
 		net_id = generate_net_id(src)
-
+	secondary_tick = rand(0, 3)
```
Comment (MarkNais) [inline code/obj/machinery/firealarm.dm:37]: I don't like this, rubs me the wrong way.

File: code/obj/machinery/firealarm.dm [hunk 5]
```diff
@@ -52,27 +53,27 @@
...
 
 /obj/machinery/firealarm/bullet_act(BLAH)
-	return src.alarm()
+	return src.alarm(5)
 
 /obj/machinery/firealarm/emp_act()
 	..()
 	if(prob(50))
-		src.alarm()
+		src.alarm(1)
```
Comment (MarkNais) [inline code/obj/machinery/firealarm.dm:76]: Magic numbers bad. I suggest putting some defines in the _std folder. Maybe like `#DEFINE ALARM_FIRE 1` etc

File: code/obj/machinery/firealarm.dm [hunk 6]
```diff
@@ -92,8 +93,19 @@
...
+		secondary_tick = 0
+		var/turf/location = src.loc
+		var/datum/gas_mixture/environment = location.return_air()
+		var/gaspressure = MIXTURE_PRESSURE(environment)
+		if(gaspressure < ONE_ATMOSPHERE*0.5)
+			src.alarm(2)
+		if(environment.toxins > 5)
+			src.alarm(3)
+		if(location.active_liquid && location.active_liquid.group && location.active_liquid.group.last_depth_level > 3)
+			src.alarm(4)
```
Comment (MarkNais) [inline code/obj/machinery/firealarm.dm:108]: See magic numbers comment

File: code/obj/machinery/firealarm.dm [hunk 7]
```diff
@@ -92,8 +93,19 @@
 /obj/machinery/firealarm/process()
 	if(status & (NOPOWER|BROKEN))
 		return
-
 	use_power(10, ENVIRON)
+	secondary_tick++
+	if(secondary_tick > 3)
```
Comment (MarkNais) [inline code/obj/machinery/firealarm.dm:98]: I recall Naksu working on variable machine ticking. Might be us both looking into.

**Review by MarkNais — CHANGES_REQUESTED** [review]

**Comment by Flaborized** [comment]
I feel like one of the most frustrating parts of this on Oshan was how you would get trapped in a room with no crowbar that was full of water, you'd press the fire alarm but then the firelocks would close instantly before you could leave cuz the water level was still super high. IMO firelocks are actually more harmful than good to everyone involved almost every time they trigger during a real fire, and I think that that's very funny but also this would be real frustrating and Not Helpful. also I could have sworn we had an unused sprite for Blue Firelocks that said O2 on them, but I'm not 100% sure

File: code/obj/machinery/firealarm.dm (hunk 7 above)
Comment (ZeWaka) [reply code/obj/machinery/firealarm.dm:98]: Yep, you just set a var.

File: code/mob/living/silicon/ai.dm (hunk 1 above)
Comment (F-OS) [reply code/mob/living/silicon/ai.dm:694]: It is, it's just very stupid. This line of code is making it so all unique alarms are classified as fire alarms for the purposes of the list index. IIRC this index also tracks motion-detecting cameras.

File: code/mob/living/silicon/robot.dm (hunk 2 above)
Comment (F-OS) [reply code/mob/living/silicon/robot.dm:974]: See previous

File: code/obj/machinery/firealarm.dm (hunk 6 above)
Comment (F-OS) [reply code/obj/machinery/firealarm.dm]: fixed

File: code/obj/machinery/firealarm.dm (hunk 4 above)
Comment (F-OS) [reply code/obj/machinery/firealarm.dm]: I feel like this is the simplest and most straightforward way to do this? Do you have any suggestions on how I could change it around to be less janky?

File: code/obj/machinery/computer/atmosphere.dm (hunk 3 above)
Comment (F-OS) [reply code/obj/machinery/computer/atmosphere.dm]: Resolving this as the code meaning is a bit different in the new commit

File: code/obj/machinery/firealarm.dm (hunk 4 above)
Comment (ZeWaka) [reply code/obj/machinery/firealarm.dm]: using the variable machine loop speed subscription service

Comment (F-OS) [reply code/obj/machinery/firealarm.dm]: Could you send me either an example or just the proc I need to call? I'm a bit confused on how to do that.

File: goonstation.dme [hunk 8]
```diff
@@ -25,6 +25,7 @@ var/datum/preMapLoad/preMapLoad = new
 #include "code\__secret_public.dme"
 
 // BEGIN_INCLUDE
+#include "_std\defines\alarms.dm"
```
Comment (ZeWaka) [inline goonstation.dme:28]: wrong place.

File: code/obj/machinery/firealarm.dm [hunk 9]
```diff
@@ -165,25 +175,28 @@
...
+/obj/machinery/firealarm/proc/post_alert(var/alarm, var/specific_target, var/type = 1)
 //	var/datum/radio_frequency/frequency = radio_controller.return_frequency(alarm_frequency)
 
 	LAGCHECK(LAG_LOW)
 
 	if(!frequency) return
 
+	var/typestring = list("Fire", "Low Pressure", "Flammable Atmosphere", "Flood", "Manual Trip")
+	if(type == 0)
+		type = 5
```
Comment (ZeWaka) [inline code/obj/machinery/firealarm.dm:187]: what

File: code/mob/living/silicon/ai.dm (hunk 1 above)
Comment (ZeWaka) [reply code/mob/living/silicon/ai.dm:694]: shitcode

File: code/obj/machinery/firealarm.dm [hunk 10]
```diff
@@ -112,10 +120,12 @@
...
 
 	if (src.icon_state == "fire0")
-		src.alarm()
+		src.alarm(ALARM_MANUALTRIP)
 	else
-		idle_count = manual_off_reactivate_idle
 		src.reset()
+		detecting = 0
+		SPAWN_DBG(manual_off_reactivate_idle)
+			detecting = 1
```
Comment (ZeWaka) [inline code/obj/machinery/firealarm.dm:128]: wat is this, a cooldown? ON_COOLDOWN is preferred for that

File: code/obj/machinery/firealarm.dm (hunk 9 above)
Comment (ZeWaka) [reply code/obj/machinery/firealarm.dm:187]: this looks like it'll runtime

**Review by ZeWaka — CHANGES_REQUESTED** [review]

**Comment by F-OS** [comment]
I'll clean this up later, too much other stuff on my plate

**Pull request closed without merging**